[InstSimplify] Optimize maximumnum and minimumnum

LewisCrawford opened the pull request:
Add support for the new maximumnum and minimumnum intrinsics in various optimizations in InstSimplify.

Also, change the behavior of optimizing maxnum(sNaN, x) to simplify to qNaN instead of x to better match the LLVM IR spec, and add more tests for sNaN behavior for all 3 max/min intrinsic types.

arsenm commented on llvm/test/Transforms/InstSimplify/fminmax-folds.ll:
```diff
@@ -521,115 +785,99 @@ define <2 x double> @minnum_nan_op1_vec(<2 x double> %x) {
-define float @maxnum_undef_op0(float %x) {
-; CHECK-LABEL: @maxnum_undef_op0(
-; CHECK-NEXT:    ret float [[X:%.*]]
+define double @maxnum_snan_op0(double %x) {
```
Unrelated test changes should be separately committed

LewisCrawford replied:
I've un-deleted the undef tests for maxnum/maximum now. I'll just not add any new undef tests for maximumnum (even thought that means the list of tests is slightly inconsistent between the 3 types of intrinsic).

arsenm commented on llvm/test/Transforms/InstSimplify/fminmax-folds.ll:
```diff
@@ -921,89 +1170,121 @@ define <2 x double> @minimum_nan_op1_vec(<2 x double> %x) {
+; CHECK-LABEL: @minimum_poison_op0(
+; CHECK-NEXT:    ret float [[X:%.*]]
+;
+  %val = call float @llvm.minimum.f32(float poison, float %x)
```
Do we have the corresponding undef changes? Also test poison, poison for good measure

LewisCrawford replied:
Yes, all versions of the intrinsics will match both undef and poison. I've avoided adding undef versions of the new tests for maximumnum due to the undef depracation effort, and all the automated checking trying to prevent new tests using undef. Initially, I also deleted the pre-existing tests for maxnum/maximum with undef too, but I've un-deleted those now.

LewisCrawford committed:
[InstSimplify] Optimize maximumnum and minimumnum

Add support for the new maximumnum and minimumnum intrinsics in various optimizations in InstSimplify.

Also, change the behavior of optimizing maxnum(sNaN, x) to simplify to qNaN instead of x to better match the LLVM IR spec, and add more tests for sNaN behavior for all 3 max/min intrinsic types.

LewisCrawford committed:
Add more vector tests + cleanup

- Add tests for vectors of <snan, qnan> pairs.
- Add more vector tests for <qnan, poison> and <snan, poison>
- Remove comment wording about FTZ.
- Clarify behavior of fmaxnum(x, <snan, qnan>) -> x in comments (assume snan == qnan)

LewisCrawford committed:
Add vector tests for all inf/max tests

Add a vec2 variant for all the +inf/-inf/+max/-max
tests for maxnum/maximum/maximumnum.

LewisCrawford commented on llvm/test/Transforms/InstSimplify/fminmax-folds.ll:
```diff
@@ -88,6 +152,23 @@ define float @test_minimum_const_inf(float %x) {
+  %r = call float @llvm.minimumnum.f32(float %x, float 0x7ff0000000000000)
+  ret float %r
+}
+
```
There was only 1 pre-existing test for vectors there before. I've added a new commit adding vector variants of all of these inf/max tests for all the intrinsics now.

LewisCrawford commented on llvm/test/Transforms/InstSimplify/fminmax-folds.ll:
```diff
@@ -218,6 +348,22 @@ define <2 x float> @test_minimum_const_inf_nnan_comm_vec(<2 x float> %x) {
+  %r = call nnan <2 x float> @llvm.minimumnum.v2f32(<2 x float> <float 0x7ff0000000000000, float 0x7ff0000000000000>, <2 x float> %x)
+  ret <2 x float> %r
+}
+
```
Done.
fmax(<x, y>, <sNaN, poison>) returns:
- <qNaN, poison> for maxnum
- <qNaN, poison> for maximum
- <x, y> for maximumnum

fmax(<x, y>, <sNaN, qNaN>) returns:
- <x, y> for maxnum (surprising, but technically legal, as we are allowed to tread sNaN as qNaN for optimizations)
- <qNaN, qNaN> for maximum
- <x, y> for maximumnum

LewisCrawford commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6711,7 +6712,16 @@ Value *llvm::simplifyBinaryIntrinsic(Intrinsic::ID IID, Type *ReturnType,
+    //
+    // For instance, we often return one of the arguments unmodified instead of
+    // inserting an llvm.canonicalize to transform input sNaNs into qNaNs or to
+    // respect any FTZ semantics, and sometimes assume all NaN inputs are qNaNs.
```
Removed part about FTZ.

LewisCrawford commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6725,32 +6735,43 @@ Value *llvm::simplifyBinaryIntrinsic(Intrinsic::ID IID, Type *ReturnType,
+    // maximum(X, nan) -> qnan
+    if (match(Op1, m_NaN())) {
+      if (PropagateNaN || (PropagateSNaN && match(Op1, m_sNaN())))
+        return propagateNaN(cast<Constant>(Op1));
```
I've rewritten this a little to make it more readable now.
The <sNaN, qNaN> handling is correct, but slightly surprising. <sNaN, qNaN> does not match with m_sNaN(), as it contains an element that is neither sNaN nor undef/poison. In this case, it will treat the incoming sNaN element as if it were a qNaN, and will return:
<x, y> for maxnum (surprising, but technically legal, as we are allowed to tread sNaN as qNaN for optimizations)
<qNaN, qNaN> for maximum
<x, y> for maximumnum
This is legal because of this wording in the langref:
> Floating-point math operations are allowed to treat all NaNs as if they were quiet NaNs. For example, “pow(1.0, SNaN)” may be simplified to 1.0.

I've added tests for this, and updated the comments in the code to explicitly state this now.

arsenm commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6741,15 +6741,14 @@ Value *llvm::simplifyBinaryIntrinsic(Intrinsic::ID IID, Type *ReturnType,
     // maxnum(x, qnan) -> x
-    // minnum(x, snan) -> qnan
-    // maxnum(x, snan) -> qnan
+    // minnum(x, snan) -> qnan (or x for vectors mixing snans and qnans)
+    // maxnum(x, snan) -> qnan (or x for vectors mixing snans and qnans)
```
This is just documenting a bug, you can't rely on the simple matcher. Either use m_APFloat and check if it's a splat quiet or signaling nan, or you have to do the complicated handling to process this per element

LewisCrawford replied:
As far as I can tell, this is still legal (if slightly surprising) behaviour because of this wording in the spec:
> Floating-point math operations are allowed to treat all NaNs as if they were quiet NaNs. For example, “pow(1.0, SNaN)” may be simplified to 1.0.

What would your preferred behaviour here be for the <sNaN, qNaN> case and <sNaN, poison> case? The way the langref states it, it seems like the preferred approach is to sacrifice sNaN correctness any time it allows an optimization (and only try to preserve it when using StrictFP or constrained FP intrinsics).

arsenm replied:
minnum/maxnum are a special case and require getting the snan handling right, because they have inverted behavior with an snan input.

but in general it's a better QoI to get the correct handling and quiet the returned value. The dropped canonicalization is more useful for instruction optimizations that may lose a quiet. In the constant folding case we can handle it elementwise, it's just more of a hassle

LewisCrawford committed:
Avoid optimization for maxnum(x, <sNaN, qNaN>)

Instead of transforming maxnum(x, <sNaN, qNaN>) -> x,
avoid folding the maxnum away so the sNaN element can
correctly be propagated as a qNaN.

LewisCrawford committed:
Update failing AMDGPU tests

Update some failing AMDGPU tests using maxnum's updated sNaN behaviour
and maximumnum's new set of optimizations.

arsenm commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6725,32 +6735,50 @@ Value *llvm::simplifyBinaryIntrinsic(Intrinsic::ID IID, Type *ReturnType,
+    // maximum(X, nan) -> qnan
+    if (PropagateSNaN && match(Op1, m_sNaN())) {
+      return propagateNaN(cast<Constant>(Op1));
+    } else if (match(Op1, m_NaN())) {
```
No else after return

LewisCrawford replied:
Done.

arsenm commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6725,32 +6735,50 @@ Value *llvm::simplifyBinaryIntrinsic(Intrinsic::ID IID, Type *ReturnType,
+    } else if (match(Op1, m_NaN())) {
+      if (PropagateNaN)
+        return propagateNaN(cast<Constant>(Op1));
+      // In cases like mixed <sNaN, qNaN> vectors, avoid the optimization to
+      // allow correct sNaN propagation where necessary.
+      else if (PropagateSNaN && !match(Op1, m_qNaN()))
+        break;
+      else
+        return Op0;
```
With this chain of nan-matchers, it might actually be easier to just do the elementwise handling - propgqtenan already has the elementwise code in it

LewisCrawford replied:
I think element-wise comparisons will end up taking this slightly out of the current scope of the patch, as it will introduce cases like: maxnum(<x, y>, <sNaN, qNaN>) -> <qNaN, y> , which will require insert/extract chains to be created, and might end up with cases where there are not clear perf wins from re-creating partially applied vectors where some elements are NaN vs just using maxnum on the original vector despite some elements being NaN. To avoid these partially applied cases, we'd need to check whether the whole vec is sNaN or qNaN (ignoring poison/undef elements), which is what the current patch does with the matchers already.

It might be worth considering these kind of element-wise folds in future (and for partial Inf, partial poison, partial FMAX vectors too in addition to NaNs maybe), but I think that would be better-suited to a separate patch.

LewisCrawford replied:
@arsenm would you be able to take a look at the new elementwise code and see if it looks ok to merge soon?

LewisCrawford committed:
Tidy up if/else chain

LewisCrawford committed:
Rewrite fmax/fmin InstSimplify elementwise

Rewrite all the fmin/fmax(x, const) optimizations in
InstSimplify to use elementwise checking instead of
pattern-matching.

This allows for improved splat-vector handling, as well as
handling of mixed constant vectors e.g. <sNan, Inf, poison>.

LewisCrawford committed:
Remove unused pattern match functions

Remove sNaN + qNaN patterns added in earlier version of
the patch, since they are no longer used in the new
elementwise approach.

LewisCrawford committed:
Fix AMDGPU tests

Re-generate 2 AMDGPU tests that are failing after rebasing this change:
 - llvm/test/CodeGen/AMDGPU/fneg-combines.new.ll
 - llvm/test/Transforms/InstCombine/AMDGPU/fmed3.ll

LewisCrawford committed:
Avoid elementwise testing for failed splats

Put the elementwise checks for fixed-vector elements in an else block
after checking for splats so that fixed-vector splats do not get
checked twice if they fail to be optimized in the first check.

LewisCrawford committed:
Fix comment consistency

Swap min/max lines to be in consistent order with other similar
comments.

LewisCrawford committed:
Merge branch 'main' into instsimplify_maximumnum

LewisCrawford committed:
Refactor to simplify

Make the code simpler, shorter, and clearer by:
 * Splitting the lambda out into a separate function
 * Instead of returning an optional of a bool + value, use an enum +
   value
 * The new enum specifies whether to avoid optimizing, return LHS,
   return const, or return either.
 * Use the same enum to cover the logic of both the individual scalar
   elements, and the vector as a whole.
 * Use this enum to cover the fact that undef/poison could choose either
   side, rather than tracking it with a separate check.
 * Extract common code from the splat vs fixed vs scalar paths.
 * Merge undef handling with the above approach too, rather than using a
   separate check.
 * Simplify some comments.

LewisCrawford commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6715,36 +6725,138 @@ Value *llvm::simplifyBinaryIntrinsic(Intrinsic::ID IID, Type *ReturnType,
+
+          bool NeedsConstElement = false;
+          bool NeedsLHSElement = false;
+          for (unsigned i = 0; i != NumElts; ++i) {
```
Hopefully the new simplified version improves this a bit, and draws out as much common code between the scalable + scalar + fixed-vec cases as possible.

LewisCrawford commented:
> So, the practical impact is that the code that currently uses `nvvm_fmin/fmax` will end up always propagating sNaNs, while the actual `fmax/fmin` PTX instruction would not.

Yes, although the sNaN propagation issue could only happen in the middle-end, since nvvm_fmax currently becomes PTX max, so it's only during the middle-end time when it is interpreted as llvm.maxnum that there could be any semantic differences (e.g. during constant-folding).

Also, this InstSimplify optimization I'm updating in this patch did not handle maxnum(x, sNaN) -> qNaN before (it used to only do maxnum(x, NaN) -> x in the same way PTX max does), so this different behaviour might become slightly more apparent after the sNaN correctness fixes included in my patch here.

> nvvm.fmin/fmax do not appear to be widely used

They are widely used in NVCC.

> Adding lowering of maximumnum/minimumnum -> fmax.f32/fmin.f32 should be pretty straightforward, too.

Yes, I'm working on a patch for this at the moment, which I should hopefully be able to upstream soon. It's 5 lines of code plus some tests, so was very straightforward to add.

LewisCrawford commented:
@arsenm Does the latest version of the patch look ok to merge?

LewisCrawford commented:
@Artem-B does this patch seem ok to merge?

Artem-B commented:
It's OK with me, but the remaining open questions are between you and @arsenm. If he's OK with the current state, you're good to go.

Artem-B reviewed (approved):
LGTM, but please check with @arsenm if all concerns had been addressed.

LewisCrawford committed:
Merge remote-tracking branch 'origin/main' into instsimplify_maximumnum

Merge branch 'main' into instsimplify_maximumnum

Replace the old fminmax-folds.ll test with the newly refactored version
from #160504. These tests will not pass yet with the new branch, but
this provides a clean baseline.

Conflicts:
        llvm/test/Transforms/InstSimplify/fminmax-folds.ll

LewisCrawford committed:
Regenerate test results

Run update_test_checks.py on fminmax-folds.ll to get the
results that change from this patchset (especially the snan/qnan
changes, and adding maximumnum/minimumnum.

LewisCrawford committed:
Update test comments (remove TODOs)

LewisCrawford committed:
Rename some CHECK variables for consistentcy

The update_test_checks.py script seems to have used the old
variable names from the pervious version of the tests, rather
than updating them to match the variable names they now represent.

Changed several nested min/max tests to use names like
MAXIMUMNUM_XY instead of MAXIMUMNUM_NESTED to make this consistent
with the other CHECK lines and avoid confusion with the nested
value that the variable used to represent in the old version of the
test.

LewisCrawford committed:
Add new mixed vector element tests

Add test-cases for the newly added elementwise optimizations
covering vectors like <poison, Inf> and <poison, Inf, SNaN>.

LewisCrawford committed:
Fix more variable naming inconsistencies

Fix cases where variable X was being renamed MINIMUMNUM

LewisCrawford committed:
Fix more variable naming issues

Fix problem with MINNUM_RES getting renamed to MINIMUM_RES

LewisCrawford committed:
Merge branch 'main' into instsimplify_maximumnum

LewisCrawford commented:
@arsenm I've updated all the fminmax-folds.ll tests now after the refactor in https://github.com/llvm/llvm-project/pull/160504

The diff is now `139 insertions(+), 106 deletions(-)`, so it should hopefully be a lot easier to review than the previous version with +1800 new lines if you've got time to take another look at this patch some times soon.

LewisCrawford committed:
Fix unused variable error

LewisCrawford committed:
Merge branch 'main' into instsimplify_maximumnum

LewisCrawford commented:
Since it's been 7 weeks since @arsenm 's last response on this bug on Aug 19th, and it's been approved for 2 weeks already, I'll just merge this patch now to avoid it getting stuck in limbo. If there are still any concerns that need addressed, I'm happy to fix those in a follow-up patch.
